File: backend/app/engines/discover.py
"""Shop many A→C itineraries, then price every layover B, and persist inversions."""

from __future__ import annotations

import argparse
import asyncio
from collections import defaultdict

import httpx

from app.config import get_settings
from app.db import repo
from app.db.session import init_db, session_factory
from app.engines.risk import attach_risk
from app.engines.shop import _priced_in_currency, _shop_providers, layover_minutes
from app.models import Offer
from app.providers.amadeus import AmadeusProvider
from app.providers.base import ShopRequest
from app.providers.bookers import booker_links
from app.providers.duffel import DuffelProvider

# Duffel test documents these as returning shaped offers (connections / stops).
SEED_PAIRS = (
    ("LHR", "DXB"),
    ("DXB", "AMS"),
    ("LHR", "JFK"),
    ("JFK", "LHR"),
    ("LHR", "CDG"),
    ("CDG", "JFK"),
    ("CDG", "LHR"),
    ("AMS", "JFK"),
    ("FRA", "JFK"),
    ("MAD", "JFK"),
    ("FCO", "JFK"),
    ("DUB", "JFK"),
    ("BOS", "LHR"),
    ("ORD", "LHR"),
    ("LAX", "LHR"),
    ("SFO", "LHR"),
    ("MIA", "LHR"),
    ("ATL", "LHR"),
    ("DFW", "LHR"),
    ("EWR", "LHR"),
    ("LGW", "JFK"),
    ("MAN", "JFK"),
    ("BCN", "JFK"),
    ("ZRH", "JFK"),
    ("VIE", "JFK"),
    ("CPH", "JFK"),
    ("OSL", "JFK"),
    ("ARN", "JFK"),
    ("LIS", "JFK"),
    ("ATH", "JFK"),
    ("IST", "LHR"),
    ("DXB", "LHR"),
    ("DOH", "LHR"),
    ("SIN", "LHR"),
    ("HKG", "LHR"),
    ("NRT", "LHR"),
    ("HND", "LAX"),
    ("SYD", "LAX"),
    ("YYZ", "LHR"),
    ("YVR", "LHR"),
    ("MEX", "JFK"),
    ("GRU", "JFK"),
    ("BTS", "MRU"),
)

HUBS = (
    "LHR", "LGW", "CDG", "AMS", "FRA", "MAD", "FCO", "DUB", "MUC", "ZRH",
    "JFK", "EWR", "BOS", "ORD", "ATL", "DFW", "LAX", "SFO", "MIA", "IAD",
    "DXB", "DOH", "IST", "SIN", "HKG", "NRT", "SYD", "YYZ",
)

SKIP_DESTS = {"STN"}  # Duffel test: STN→LHR is a forced timeout
DEFAULT_DATES = ("2026-10-22", "2026-11-05", "2026-11-19")


def _cheapest_to(offers: list[Offer], dest: str) -> Offer | None:
    priced = [
        o
        for o in offers
        if o.price is not None and o.segments and o.segments[-1].dest.upper() == dest.upper()
    ]
    priced = _priced_in_currency(priced)
    if not priced:
        return None
    return min(priced, key=lambda o: (o.price or 1e12, layover_minutes(o), o.duration_min))


async def _shop(
    req: ShopRequest,
    amadeus: AmadeusProvider | None,
    duffel: DuffelProvider | None,
    sem: asyncio.Semaphore,
) -> list[Offer]:
    async with sem:
        try:
            offers = await _shop_providers(req, amadeus, duffel)
        except Exception as exc:
            print(f"  shop error {req.origin}->{req.dest} {req.date}: {exc}", flush=True)
            return []
        await asyncio.sleep(0.35)
        return offers


async def _dests_for(origin: str, extra: list[str], limit: int) -> list[str]:
    async with session_factory()() as session:
        spokes = [code for code, _, _ in await repo.destinations_from(session, origin, limit=limit)]
    out: list[str] = []
    seen = {origin.upper(), *SKIP_DESTS}
    for code in [*extra, *spokes]:
        c = code.upper()
        if len(c) != 3 or c in seen:
            continue
        seen.add(c)
        out.append(c)
    return out


async def discover_hidden_deals(
    *,
    limit: int = 28,
    dests_per_origin: int = 22,
    day: str | None = None,
    dates: list[str] | None = None,
) -> dict:
    settings = get_settings()
    await init_db()
    days = [day] if day else list(dates or DEFAULT_DATES)
    if not settings.amadeus_enabled and not settings.duffel_enabled:
        return {"scanned": 0, "found_this_run": 0, "stored": 0, "errors": ["No shop keys"], "dates": days}

    found = 0
    scanned = 0
    shops = 0
    connecting_seen = 0
    compared = 0
    empty_shops = 0
    closest: list[tuple[float, str]] = []
    errors: list[str] = []
    hubs = list(HUBS[:limit])
    extras_by_origin: dict[str, list[str]] = defaultdict(list)
    for o, d in SEED_PAIRS:
        extras_by_origin[o].append(d)
        if o not in hubs:
            hubs.append(o)

    sem = asyncio.Semaphore(max(2, min(settings.max_concurrency, 3)))

    async with httpx.AsyncClient(timeout=httpx.Timeout(45.0, connect=10.0)) as client:
        amadeus = AmadeusProvider(settings, client) if settings.amadeus_enabled else None
        duffel = DuffelProvider(settings, client) if settings.duffel_enabled else None
        for date in days:
            print(f"\n===== {date} =====", flush=True)
            for origin in hubs:
                dests = await _dests_for(origin, extras_by_origin.get(origin, []), dests_per_origin)
                if not dests:
                    continue
                scanned += 1
                print(f"{origin} {date}: shopping {len(dests)} destinations", flush=True)
                by_dest: dict[str, list[Offer]] = {}
                all_offers: list[Offer] = []
                for dest in dests:
                    req = ShopRequest(origin=origin, dest=dest, date=date, nonstop=False, max_offers=25)
                    offers = await _shop(req, amadeus, duffel, sem)
                    shops += 1
                    if not offers:
                        empty_shops += 1
                        continue
                    by_dest[dest] = offers
                    all_offers.extend(offers)

                missing_b: set[str] = set()
                for offer in all_offers:
                    if offer.price is None or len(offer.segments) < 2:
                        continue
                    connecting_seen += 1
                    dest_b = offer.segments[0].dest.upper()
                    dest_c = offer.segments[-1].dest.upper()
                    if dest_b in {origin, dest_c} or dest_b in SKIP_DESTS:
                        continue
                    if dest_b not in by_dest:
                        missing_b.add(dest_b)

                for dest_b in sorted(missing_b)[:20]:
                    req = ShopRequest(origin=origin, dest=dest_b, date=date, nonstop=False, max_offers=25)
                    offers = await _shop(req, amadeus, duffel, sem)
                    shops += 1
                    if offers:
                        by_dest[dest_b] = offers
                        all_offers.extend(offers)
                    else:
                        empty_shops += 1

                cheapest_b = {d: _cheapest_to(offers, d) for d, offers in by_dest.items()}
                matches = []
                seen: set[tuple[str, str, str]] = set()
                async with session_factory()() as session:
                    o_ap = await repo.get_airport(session, origin)
                    if not o_ap:
                        errors.append(f"{origin}: unknown")
                        continue
                    for offer in all_offers:
                        if offer.price is None or len(offer.segments) < 2:
                            continue
                        dest_b = offer.segments[0].dest.upper()
                        dest_c = offer.segments[-1].dest.upper()
                        if dest_b in {origin, dest_c}:
                            continue
                        local = cheapest_b.get(dest_b)
                        if local is None or local.price is None or offer.currency != local.currency:
                            continue
                        compared += 1
                        if offer.price >= local.price:
                            gap = (offer.price / local.price) if local.price else 99
                            closest.append(
                                (
                                    gap,
                                    f"{origin}->{dest_b}->{dest_c} through={offer.price} local={local.price} {offer.currency}",
                                )
                            )
                            continue
                        key = (origin, dest_b, dest_c, date)
                        if key in seen:
                            continue
                        seen.add(key)
                        b_ap = await repo.get_airport(session, dest_b)
                        c_ap = await repo.get_airport(session, dest_c)
                        if not b_ap or not c_ap:
                            continue
                        match = attach_risk(
                            match_id=f"disc-{origin}-{dest_b}-{dest_c}-{offer.id}",
                            hidden_city=dest_c,
                            hidden_name=f"{c_ap.city} ({dest_c})",
                            local=local,
                            through=offer,
                            true_dest=dest_b,
                            origin_country=o_ap.country,
                            hidden_country=c_ap.country,
                        )
                        match.bookers = booker_links(origin, dest_c, date, 1, currency=offer.currency)
                        await repo.persist_hidden_deals(
                            session,
                            [match],
                            origin=origin,
                            origin_city=o_ap.city,
                            dest=dest_b,
                            dest_city=b_ap.city,
                            date=date,
                        )
                        matches.append(match)
                        print(
                            f"  HIT {origin}->{dest_b}->{dest_c} through={offer.price} {offer.currency} "
                            f"local={local.price} save={match.gross_saving}",
                            flush=True,
                        )
                found += len(matches)
                print(
                    f"{origin} {date}: {len(matches)} inversion(s), {len(all_offers)} offers, "
                    f"{sum(1 for o in all_offers if len(o.segments) > 1)} connecting",
                    flush=True,
                )

    async with session_factory()() as session:
        deals = await repo.list_hidden_deals(session, limit=120)
    summary = {
        "scanned": scanned,
        "shops": shops,
        "empty_shops": empty_shops,
        "connecting_seen": connecting_seen,
        "compared_cheaper": compared,
        "found_this_run": found,
        "stored": len(deals),
        "errors": errors,
        "dates": days,
        "closest_misses": [row for _gap, row in sorted(closest)[:8]],
    }
    print(summary, flush=True)
    return summary


async def main() -> None:
    parser = argparse.ArgumentParser(description="Scan shop APIs for hidden-city inversions")
    parser.add_argument("--limit", type=int, default=28)
    parser.add_argument("--dests", type=int, default=22)
    parser.add_argument("--date", default="")
    args = parser.parse_args()
    await discover_hidden_deals(
        limit=args.limit,
        dests_per_origin=args.dests,
        day=args.date or None,
    )


if __name__ == "__main__":
    asyncio.run(main())

File: backend/src/engines/discover.rs
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use std::time::Duration;

use chrono::Local;
use serde_json::{json, Value};
use sqlx::PgPool;
use tokio::sync::Semaphore;

use crate::config::Settings;
use crate::db::repo;
use crate::engines::hidden::{meaningful_saving, ticketed_destination};
use crate::engines::risk::attach_risk;
use crate::engines::shop::{_drop_sandbox, _priced_in_currency, _shop_providers, layover_minutes};
use crate::fx::{offer_in_usd, refresh_rates};
use crate::models::{Offer, ShopRequest};
use crate::providers::bookers::booker_links;
use crate::providers::duffel::DuffelProvider;
use crate::providers::sandbox::is_live_fare;

pub fn _scan_dates() -> Vec<String> {
    let today = Local::now().date_naive();
    [21i64, 35, 49]
        .into_iter()
        .map(|offset| (today + chrono::Duration::days(offset)).format("%Y-%m-%d").to_string())
        .collect()
}

pub fn _via_stops(offer: &Offer, origin: &str) -> (String, Vec<(i32, String)>) {
    let ticketed = ticketed_destination(offer);
    let mut idx = offer
        .outbound_end
        .unwrap_or(offer.segments.len().saturating_sub(1) as i32);
    if !offer.segments.is_empty() {
        idx = idx.max(0).min((offer.segments.len() - 1) as i32);
    }
    let origin_u = origin.to_uppercase();
    let mut vias = Vec::new();
    let end = if offer.segments.is_empty() {
        0
    } else {
        idx as usize
    };
    for (i, seg) in offer.segments.iter().take(end).enumerate() {
        let dest_b = seg.dest.to_uppercase();
        if dest_b != origin_u && dest_b != ticketed {
            vias.push((i as i32, dest_b));
        }
    }
    (ticketed, vias)
}

pub fn _cheapest_to(offers: &[Offer], dest: &str) -> Option<Offer> {
    let dest_u = dest.to_uppercase();
    let priced: Vec<Offer> = offers
        .iter()
        .filter(|o| {
            o.price.is_some() && !o.segments.is_empty() && ticketed_destination(o) == dest_u
        })
        .cloned()
        .collect();
    let priced = _priced_in_currency(&priced, None);
    priced.into_iter().min_by(|a, b| {
        let pa = a.price.unwrap_or(1e12);
        let pb = b.price.unwrap_or(1e12);
        pa.partial_cmp(&pb)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| layover_minutes(a).cmp(&layover_minutes(b)))
            .then_with(|| a.duration_min.cmp(&b.duration_min))
    })
}

async fn shop(
    req: &ShopRequest,
    duffel: Option<&DuffelProvider>,
    sem: &Semaphore,
) -> Vec<Offer> {
    let _g = match sem.acquire().await {
        Ok(g) => g,
        Err(_) => return vec![],
    };
    let offers = _shop_providers(req, duffel, false, None, "discover").await;
    tokio::time::sleep(Duration::from_millis(350)).await;
    _priced_in_currency(&_drop_sandbox(&offers), None)
}

async fn dests_for(
    pool: &PgPool,
    origin: &str,
    extra: &[String],
    limit: i64,
    skip: &HashSet<String>,
) -> anyhow::Result<Vec<String>> {
    let spokes: Vec<String> = repo::destinations_from(pool, origin, limit)
        .await?
        .into_iter()
        .map(|(code, _, _)| code)
        .collect();
    let mut out = Vec::new();
    let mut seen = skip.clone();
    seen.insert(origin.to_uppercase());
    for code in extra.iter().chain(spokes.iter()) {
        let c = code.to_uppercase();
        if c.len() != 3 || seen.contains(&c) {
            continue;
        }
        seen.insert(c.clone());
        out.push(c);
    }
    Ok(out)
}

pub async fn discover_hidden_deals(
    pool: &PgPool,
    settings: &Settings,
    client: &reqwest::Client,
    limit: i64,
    dests_per_origin: i64,
    day: Option<String>,
    dates: Option<Vec<String>>,
) -> anyhow::Result<Value> {
    let days = if let Some(day) = day.filter(|d| !d.is_empty()) {
        vec![day]
    } else {
        dates.unwrap_or_else(_scan_dates)
    };
    let skip = settings.skip_dests();
    refresh_rates(client).await;
    if !settings.duffel_enabled() {
        return Ok(json!({
            "scanned": 0,
            "found_this_run": 0,
            "stored": 0,
            "errors": ["No Duffel token"],
            "dates": days,
        }));
    }
    if !settings.publish_deals() {
        return Ok(json!({
            "scanned": 0,
            "found_this_run": 0,
            "stored": 0,
            "errors": ["Need a Duffel live token (duffel_live_…) to publish deals. Test inventory is not saved."],
            "dates": days,
        }));
    }

    let mut found = 0i64;
    let mut scanned = 0i64;
    let mut shops = 0i64;
    let mut connecting_seen = 0i64;
    let mut compared = 0i64;
    let mut empty_shops = 0i64;
    let mut closest: Vec<(f64, String)> = Vec::new();
    let mut errors: Vec<String> = Vec::new();
    let mut extras_by_origin: HashMap<String, Vec<String>> = HashMap::new();

    let hubs = repo::busiest_airports(pool, limit, &skip).await?;
    for o in &hubs {
        extras_by_origin.insert(
            o.clone(),
            repo::destinations_from(pool, o, 8)
                .await?
                .into_iter()
                .map(|(d, _, _)| d)
                .filter(|d| !skip.contains(&d.to_uppercase()))
                .collect(),
        );
    }

    let sem_n = 2.max(settings.max_concurrency.min(3));
    let sem = Arc::new(Semaphore::new(sem_n.max(1)));
    let duffel = if settings.duffel_enabled() {
        Some(DuffelProvider::new(settings, client.clone()))
    } else {
        None
    };

    for date in &days {
        println!("\n===== {date} =====");
        for origin in &hubs {
            let extra = extras_by_origin.get(origin).cloned().unwrap_or_default();
            let dests = dests_for(pool, origin, &extra, dests_per_origin, &skip).await?;
            if dests.is_empty() {
                continue;
            }
            scanned += 1;
            println!("{origin} {date}: shopping {} destinations", dests.len());
            let mut by_dest: HashMap<String, Vec<Offer>> = HashMap::new();
            let mut all_offers: Vec<Offer> = Vec::new();
            for dest in &dests {
                let req = ShopRequest {
                    origin: origin.clone(),
                    dest: dest.clone(),
                    date: date.clone(),
                    adults: 1,
                    cabin: "ECONOMY".into(),
                    currency: "USD".into(),
                    nonstop: false,
                    max_offers: 25,
                    return_date: None,
                };
                let offers = shop(&req, duffel.as_ref(), sem.as_ref()).await;
                shops += 1;
                if offers.is_empty() {
                    empty_shops += 1;
                    continue;
                }
                by_dest.insert(dest.clone(), offers.clone());
                all_offers.extend(offers);
            }

            let mut missing_b: HashSet<String> = HashSet::new();
            for offer in &all_offers {
                if offer.price.is_none() || offer.segments.len() < 2 {
                    continue;
                }
                connecting_seen += 1;
                let (_ticketed, vias) = _via_stops(offer, origin);
                for (_i, dest_b) in vias {
                    if skip.contains(&dest_b) {
                        continue;
                    }
                    if !by_dest.contains_key(&dest_b) {
                        missing_b.insert(dest_b);
                    }
                }
            }

            let mut missing: Vec<String> = missing_b.into_iter().collect();
            missing.sort();
            for dest_b in missing.into_iter().take(20) {
                let req = ShopRequest {
                    origin: origin.clone(),
                    dest: dest_b.clone(),
                    date: date.clone(),
                    adults: 1,
                    cabin: "ECONOMY".into(),
                    currency: "USD".into(),
                    nonstop: false,
                    max_offers: 25,
                    return_date: None,
                };
                let offers = shop(&req, duffel.as_ref(), sem.as_ref()).await;
                shops += 1;
                if !offers.is_empty() {
                    by_dest.insert(dest_b, offers.clone());
                    all_offers.extend(offers);
                } else {
                    empty_shops += 1;
                }
            }

            let cheapest_b: HashMap<String, Offer> = by_dest
                .iter()
                .filter_map(|(d, offers)| _cheapest_to(offers, d).map(|o| (d.clone(), o)))
                .collect();
            let mut matches = Vec::new();
            let mut seen: HashSet<(String, String, String, String)> = HashSet::new();
            let Some(o_ap) = repo::get_airport(pool, origin, false).await? else {
                errors.push(format!("{origin}: unknown"));
                continue;
            };
            for offer in &all_offers {
                if offer.price.is_none() || offer.segments.len() < 2 {
                    continue;
                }
                let (dest_c, vias) = _via_stops(offer, origin);
                if dest_c.is_empty() {
                    continue;
                }
                for (exit_i, dest_b) in vias {
                    let Some(local) = cheapest_b.get(&dest_b) else {
                        continue;
                    };
                    if local.price.is_none() {
                        continue;
                    }
                    if !is_live_fare(offer) || !is_live_fare(local) {
                        continue;
                    }
                    let Some(through_usd) = offer_in_usd(offer) else {
                        continue;
                    };
                    let Some(local_usd) = offer_in_usd(local) else {
                        continue;
                    };
                    let (Some(tp), Some(lp)) = (through_usd.price, local_usd.price) else {
                        continue;
                    };
                    compared += 1;
                    if tp >= lp {
                        let gap = if lp != 0.0 { tp / lp } else { 99.0 };
                        closest.push((
                            gap,
                            format!(
                                "{origin}->{dest_b}->{dest_c} through={tp} local={lp} USD"
                            ),
                        ));
                        continue;
                    }
                    let saving = ((lp - tp) * 100.0).round() / 100.0;
                    if !meaningful_saving(Some(saving), Some(settings.min_hidden_saving)) {
                        continue;
                    }
                    let key = (
                        origin.clone(),
                        dest_b.clone(),
                        dest_c.clone(),
                        date.clone(),
                    );
                    if !seen.insert(key) {
                        continue;
                    }
                    let Some(b_ap) = repo::get_airport(pool, &dest_b, false).await? else {
                        continue;
                    };
                    let Some(c_ap) = repo::get_airport(pool, &dest_c, false).await? else {
                        continue;
                    };
                    let mut match_ = attach_risk(
                        &format!("disc-{origin}-{dest_b}-{dest_c}-{}", offer.id),
                        &dest_c,
                        &format!("{} ({dest_c})", c_ap.city),
                        &local_usd,
                        &through_usd,
                        &dest_b,
                        &o_ap.country,
                        &c_ap.country,
                        exit_i,
                    );
                    match_.bookers = booker_links(
                        origin,
                        &dest_c,
                        date,
                        1,
                        &[],
                        "USD",
                        "ECONOMY",
                        None,
                    );
                    repo::persist_hidden_deals(
                        pool,
                        &[match_.clone()],
                        origin,
                        &o_ap.city,
                        &dest_b,
                        &b_ap.city,
                        date,
                    )
                    .await?;
                    println!(
                        "  HIT {origin}->{dest_b}->{dest_c} through={tp} USD local={lp} save={saving}"
                    );
                    matches.push(match_);
                }
            }
            found += matches.len() as i64;
            let connecting_n = all_offers.iter().filter(|o| o.segments.len() > 1).count();
            println!(
                "{origin} {date}: {} inversion(s), {} offers, {connecting_n} connecting",
                matches.len(),
                all_offers.len()
            );
        }
    }

    let deals = repo::list_hidden_deals(pool, 120, "", "").await?;
    closest.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));
    let closest_misses: Vec<String> = closest.into_iter().take(8).map(|(_, row)| row).collect();
    let summary = json!({
        "scanned": scanned,
        "shops": shops,
        "empty_shops": empty_shops,
        "connecting_seen": connecting_seen,
        "compared_cheaper": compared,
        "found_this_run": found,
        "stored": deals.len(),
        "errors": errors,
        "dates": days,
        "closest_misses": closest_misses,
    });
    println!("{summary}");
    Ok(summary)
}
